File: Template/Source/combine_runs.f
      program test
c*****************************************************************
c     Program to combine results from channels that have been
c     split into multiple jobs. Multi-job channels are identified
c     by local file mjobs.dat in the channel directory.
c****************************************************************
      implicit none
c
c     Constants
c
      include 'run_config.inc'
      integer    maxsubprocesses
      parameter (maxsubprocesses=999)
c     
c     Local
c
      character*300 procname(maxsubprocesses)
      character*300 channame(maxsubprocesses)
      character*300 pathname
      integer iproc,nproc
      integer ichan,nchan
      integer jc

c-----
c  Begin Code
c-----

c
c     Get list of subprocesses
c
      call get_subprocess(procname,nproc)

      do iproc = 1, nproc   !Loop over each subprocess
         jc = index(procname(iproc),' ')
         pathname = procname(iproc)(1:jc)
c
c        get list of integration channels for this subprocess
c
         call get_channels(pathname,channame,nchan)
         do ichan = 1, nchan
            call sum_multichannel(channame(ichan))
         enddo
      enddo

      end

      subroutine get_subprocess(subname,ns)
c*****************************************************************
c     Opens file subproc.mg to determine all subprocesses
c****************************************************************
      implicit none
c
c     Constants
c
      character*(*) plist
      parameter    (plist='subproc.mg')
      integer    maxsubprocesses
      parameter (maxsubprocesses=999)
c
c     Arguments
c
      character*(*) subname(maxsubprocesses)
      integer ns
c-----
c  Begin Code
c-----
      ns = 1
      open(unit=15, file=plist,status='old',err=99)
      do while (.true.)
         read(15,*,err=999,end=999) subname(ns)
         ns=ns+1
      enddo
 99   subname(ns) = './'
      write(*,*) "Did not find ", plist
      return
 999  ns = ns-1
      write(*,*) "Found ", ns," subprocesses"
      close(15)
      end

      subroutine get_channels(pathname,channame,nchan)
c*****************************************************************
c     Opens file symfact.dat to determine all channels
c****************************************************************
      implicit none
c
c     Constants
c
      character*(*) symfile
      parameter    (symfile='symfact.dat')
      include 'maxparticles.inc'
      integer    maxsubprocesses
      parameter (maxsubprocesses=999)
c
c     Arguments
c
      character*(*) channame(maxsubprocesses)
      character*(*) pathname
      integer nchan
c
c     Local
c
      character*300 fname
      character*50 dirname
      integer jc,ip
      double precision xi
      integer j,k
      integer ncode,npos
      character*20 formstr
c-----
c  Begin Code
c-----
      jc = index(pathname," ")
c     ncode is number of digits needed for the bw coding
      ncode=int(dlog10(3d0)*(max_particles-3))+1
      fname = pathname(1:jc-1) // "/" // symfile
      nchan = 0
      open(unit=35, file=fname,status='old',err=99)
      do while (.true.)
         read(35,*,err=99,end=99) xi,j
         if (j .gt. 0) then
            k = int(xi*(1+10**-ncode))
            npos=int(dlog10(dble(k)))+1
            if ( (xi-k) .eq. 0) then
c              Write with correct number of digits
               write(formstr,'(a,i1,a)') '(a,i',npos,',a)'
               write(dirname, formstr) 'G',k,'/'
            else               !Handle B.W.
c              Write with correct number of digits
               write(formstr,'(a,i1,a,i1,a)') '(a,f',npos+ncode+1,
     $                 '.',ncode,',a)'
               write(dirname,formstr)  'G',xi,'/'
            endif     
            ip = index(dirname,'/')
            nchan=nchan+1            
            channame(nchan) = fname(1:jc-1)// "/" //dirname(1:ip)
         endif
 98   enddo
 99   close(35)
      write(*,*) "Found ", nchan, " channels."
      end

      subroutine sum_multichannel(channame)
c*****************************************************************
c     Looks in channel to see if there are multiple runs that
c     need to be combined. If so combines them into single run
c****************************************************************
      implicit none
c
c     Constants
c
      character*(*) mfile
      parameter    (mfile='multijob.dat')
      integer       maxjobs
      parameter    (maxjobs=26)
c
c     Arguments
c
      character*(*) channame
c
c     Local
c
      character*300 fname
      character*300 pathname
      character*26 alphabet
      integer jc,i
      integer njobs
      double precision xsec(maxjobs),xerr(maxjobs)
      double precision xsec_it(9,maxjobs),xerr_it(9,maxjobs)
      double precision eff(9,maxjobs),wmax(9,maxjobs)
      integer nevents(maxjobs),ntry(maxjobs)
      double precision xsec_tot, xerr_tot
      integer tot_nevents, tot_ntry
      integer lunw,nw
      double precision wgt
c-----
c  Begin Code
c-----
      alphabet="abcdefghijklmnopqrstuvwxyz"
      jc = index(channame," ")
      fname = channame(1:jc-1) // mfile
      njobs = 0
      open(unit=35, file=fname,status='old',err=99)
      read(35,*,err=99,end=99) njobs
 99   close(35)
      if(njobs .gt. 0) then
         write(*,*) "Found ", njobs, " jobs in ",fname
c
c     Read in results from each of the jobs
c
      do i = 1, njobs
         pathname = channame(1:jc-2) // alphabet(i:i)
         call get_results(pathname,xsec(i),xerr(i),ntry(i),nevents(i),
     $        xsec_it(1,i),xerr_it(1,i),eff(1,i),wmax(1,i))
      enddo
c
c     Process results from each job to get final statistics
c
      xsec_tot = 0d0
      xerr_tot = 0d0
      tot_ntry = 0
      tot_nevents = 0
      do i = 1, njobs
         xsec_tot = xsec_tot+xsec(i)
         xerr_tot = xerr_tot+xerr(i)**2
         tot_ntry = tot_ntry+ntry(i)
         tot_nevents = tot_nevents+nevents(i)
      enddo
      xsec_tot = xsec_tot / njobs
      xerr_tot = sqrt(xerr_tot)/njobs

      pathname = channame(1:jc-2)
      i=1
      call put_results(pathname,xsec_tot,xerr_tot,tot_ntry,tot_nevents,
     $        xsec_it(1,i),xerr_it(1,i),eff(1,i),wmax(1,i))
      call write_logfile(pathname,xsec,xerr,ntry,nevents,
     $        xsec_it,xerr_it,eff,wmax,njobs)
c
c     Now read in all of the events and write them
c     back out with the appropriate scaled weight
c      
      lunw=15
      jc = index(pathname," ")
      open(unit=lunw,file=pathname(1:jc-1)//"/"//"events.lhe",
     $     status="unknown",err=999)
      write(*,*) "Placing combined events in ",
     $     pathname(1:jc-1)//"/"//"events.lhe"
      wgt = xsec_tot / tot_nevents
      tot_nevents=0
      do i = 1, njobs
         jc = index(channame," ")
         pathname = channame(1:jc-2) // alphabet(i:i)
         call copy_events(pathname,lunw,wgt,nw)
         tot_nevents=tot_nevents+nw
         write(*,*) "Added ",nw," events from run ", alphabet(i:i)
         if (nw .ne. nevents(i)) then
            write(*,*) "Error writing events ",i,nw,nevents(i),
     $           pathname
         else
         endif
      enddo
      write(*,*) "Combined ",tot_nevents," to ",channame(1:jc-2)
      endif
      close(lunw)
      return
 999  close(lunw)
      write(*,*) "Error, unable to open events.lhe file for output"
     $     ,pathname
      end

      subroutine copy_events(pathname,lunw,wgt,nw)
c*********************************************************************
c     Copy events from separate runs into one file w/ appropriate wgts
c*********************************************************************
      implicit none
c     
c     Constants
c
      character*(*) eventfname
      parameter    (eventfname="events.lhe")
      include 'maxparticles.inc'
      integer    maxexternal     !Max number external momenta
      parameter (maxexternal=2*max_particles-3)
c     
c     Arguments
c
      character*(*) pathname
      double precision wgt
      integer lunw,nw
c     
c     Local
c
      character*300 fname
      integer jc
      double precision P(0:4,maxexternal),xwgt
      integer n,ic(7,maxexternal),ievent
      double precision scale,aqcd,aqed
      character*140 buff
      logical done
c-----
c  Begin Code
c-----
      jc = index(pathname," ")
      fname = pathname(1:jc-1) // "/" // eventfname
      nw = 0
      write(*,*) "Copy from ",fname(1:jc+10)
      open(unit=35, file=fname,status='old',err=999)
      done = .false.
      do while (.not. done .and. nw < 999999)
         done = .true.
         call read_event(35,P,xwgt,n,ic,ievent,scale,
     $        aqcd,aqed,buff,done)
         if (.not. done) then
            call write_event(lunw,P,wgt,n,ic,ievent,scale,
     $           aqcd,aqed,buff)
            nw=nw+1
         endif
      enddo
 99   close(35)
      return
 999  close(35)
      write(*,*) "Error unable to open ",fname
      end

      subroutine get_results(pathname,xsec,xerr,ntry,nevents,
     $     xsec_it,xerr_it,eff,wmax)
c*********************************************************************
c     Read run results from file results.dat
c*********************************************************************
      implicit none
c     
c     Constants
c
      character*(*) resultfname
      parameter    (resultfname="results.dat")
c     
c     Arguments
c
      character*(*) pathname
      double precision xsec,xerr
      double precision xsec_it(9),xerr_it(9)
      double precision eff(9),wmax(9)
      integer nevents,ntry
c     
c     Local
c
      character*300 fname
      integer jc,i,j
      double precision x1,x2,x3,x4
c-----
c  Begin Code
c-----
      jc = index(pathname," ")
      fname = pathname(1:jc-1) // "/" // resultfname
      nevents = 0
      i=1
      open(unit=35, file=fname,status='old',err=99)
      read(35,*,err=99,end=99) xsec,xerr, x1, ntry,x3,x4,nevents
      do while (.true.)
         read(35,*,end=99,err=99) x1,xsec_it(i),xerr_it(i),
     $        eff(i),wmax(i)
         i=i+1
      enddo
 99   close(35)
      i=i-1
      xsec_it(i+1) = -1d0
      if (nevents .gt. 0) then
         write(*,*) "Found ",nevents, " events in ", pathname(1:jc-1)
         i=1
         do while (xsec_it(i) .gt. 0d0)
            write(*,*) i,xsec_it(i),xerr_it(i),eff(i),wmax(i)
            i=i+1
         enddo
      endif      
      end

      subroutine write_logfile(pathname,xsec,xerr,ntry,nevents,
     $     xsec_it,xerr_it,eff,wmax,njobs)
c*********************************************************************
c     Read run results from file results.dat
c*********************************************************************
      implicit none
c     
c     Constants
c
      integer       maxjobs
      parameter    (maxjobs=26)
      character*(*) logfname
      parameter    (logfname="log.txt")
c     
c     Arguments
c
      character*(*) pathname
      double precision xsec(maxjobs),xerr(maxjobs)
      double precision xsec_it(9,maxjobs),xerr_it(9,maxjobs)
      double precision eff(9,maxjobs),wmax(9,maxjobs)
      integer nevents(maxjobs),ntry(maxjobs)
      double precision xsec_tot, xerr_tot
      integer tot_nevents, tot_ntry
      integer njobs
c     
c     Local
c
      character*300 fname
      character*26 alphabet
      integer jc,i,j
c-----
c  Begin Code
c-----
      alphabet="abcdefghijklmnopqrstuvwxyz"
      jc = index(pathname," ")
      fname = pathname(1:jc-1) // "/" // logfname
      open(unit=35,file=fname,status="old",access="append",err=999)

      write(35,*) '--------------------- Multi run with ',njobs,' jobs. ',
     $            '---------------------'
      do i=1,njobs
         write(35,*) "job ",alphabet(i:i),":",xsec(i),nevents(i)
      enddo
      close(35)
      return
 999  write(*,*) "Error opening ",fname(1:jc+10), "for append"
      close(35)
      end

      subroutine put_results(pathname,xsec,xerr,ntry,nevents,
     $     xsec_it,xerr_it,eff,wmax)
c*********************************************************************
c     Read run results from file results.dat
c*********************************************************************
      implicit none
c     
c     Constants
c
      character*(*) resultfname
      parameter    (resultfname="results.dat")
c     
c     Arguments
c
      character*(*) pathname
      double precision xsec,xerr
      double precision xsec_it(9),xerr_it(9)
      double precision eff(9),wmax(9)
      integer nevents,ntry
c     
c     Local
c
      character*300 fname
      integer jc,i,j
      double precision x1,x2
      integer i3,i4
c-----
c  Begin Code
c-----
      jc = index(pathname," ")
      fname = pathname(1:jc-1) // "/" // resultfname
      x1 = 0d0
      x2 = 0d0
      i3 = 0
      i4 = 0
      open(unit=35, file=fname,status='unknown',err=99)
      write(35,'(3e12.5,2i9,i5,i9,e10.3)') xsec,xerr, x1, ntry,i3,i4
     $     ,nevents,(1.00*nevents)/xsec
      i=1
      do while (xsec_it(i) .gt. 0 .and. i .lt. 10)
         write(35,'(i4,4e15.5)') i,xsec_it(i),xerr_it(i),
     $        eff(i),wmax(i)
         i=i+1
      enddo
 99   close(35)
      end

File: PROC_sm_14/bin/internal/combine_runs.py
################################################################################
#
# Copyright (c) 2012 The MadGraph5_aMC@NLO Development team and Contributors
#
# This file is a part of the MadGraph5_aMC@NLO project, an application which 
# automatically generates Feynman diagrams and matrix elements for arbitrary
# high-energy processes in the Standard Model and beyond.
#
# It is subject to the MadGraph5_aMC@NLO license which should accompany this 
# distribution.
#
# For more information, visit madgraph.phys.ucl.ac.be and amcatnlo.web.cern.ch
#
################################################################################
"""Program to combine results from channels that have been
     split into multiple jobs. Multi-job channels are identified
     by local file mjobs.dat in the channel directory.
"""
from __future__ import division
from __future__ import absolute_import
import math
import os
import shutil
import re
import logging
from six.moves import range
import random
import time


try:
    import madgraph
except ImportError:
    import internal.sum_html as sum_html
    import internal.misc as misc
    from internal import InvalidCmd, MadGraph5Error    
else:
    import madgraph.madevent.sum_html as sum_html
    import madgraph.various.misc as misc
    from madgraph import InvalidCmd, MadGraph5Error, MG5DIR

    
logger = logging.getLogger('madevent.combine_run') # -> stdout

#usefull shortcut
pjoin = os.path.join

   
def get_inc_file(path):
    """read the information of fortran inc files and returns
       the definition in a dictionary format.
       This catch PARAMETER (NAME = VALUE)"""
       
    pat = re.compile(r'''PARAMETER\s*\((?P<name>[_\w]*)\s*=\s*(?P<value>[\+\-\ded]*)\)''',
                     re.I)
        
    out = {}   
    for name, value in pat.findall(open(path).read()):
        orig_value = str(value)
        try:
            out[name.lower()] = float(value.replace('d','e'))
        except ValueError:
            out[name] = orig_value
    return out

class CombineRuns(object):
    
    def __init__(self, me_dir, subproc=None):
        
        self.me_dir = me_dir
        
        if not subproc:
            subproc = [l.strip() for l in open(pjoin(self.me_dir,'SubProcesses', 
                                                                 'subproc.mg'))]
        self.subproc = subproc
        maxpart = get_inc_file(pjoin(me_dir, 'Source', 'maxparticles.inc'))
        self.maxparticles = maxpart['max_particles']
    
    
        for procname in self.subproc:
            path = pjoin(self.me_dir,'SubProcesses', procname)
            channels = self.get_channels(path)
            for channel in channels:
                self.sum_multichannel(channel)
    
    def sum_multichannel(self, channel):
        """Looks in channel to see if there are multiple runs that
        need to be combined. If so combines them into single run"""
       
        start = time.time()
        alphabet = "abcdefghijklmnopqrstuvwxyz"

        if os.path.exists(pjoin(channel, 'multijob.dat')):
            njobs = int(open(pjoin(channel, 'multijob.dat')).read())
        else:
            return
        results = sum_html.Combine_results(channel)
        if njobs:
            logger.debug('find %s multijob in %s' % (njobs, channel))
        else:
            return
        for i in range(njobs):
            if channel.endswith(os.path.pathsep):
                path = channel[:-1] + alphabet[i % 26] + str((i+1)//26) 
            else:
                path = channel + alphabet[i % 26] + str((i+1)//26) 
            results.add_results(name=alphabet[i % 26] + str((i+1)//26) , 
                                filepath=pjoin(path, 'results.dat'))
        
        results.compute_average()
        if results.xsec:
            results.write_results_dat(pjoin(channel, 'results.dat'))
        else:
            return
        ### Adding information in the log file
        fsock = open(pjoin(channel, 'log.txt'), 'a')
        fsock.write('--------------------- Multi run with %s jobs. ---------------------\n'
                    % njobs)
        for r in results:
            
            fsock.write('job %s : %s %s +- %s %s\n' % (r.name, r.xsec, r.axsec,\
                                                       r.xerru, r.nunwgt))  
            
        #Now read in all of the events and write them
        #back out with the appropriate scaled weight
        to_clean = []
        fsock = open(pjoin(channel, 'events.lhe'), 'w')
        #wgt = results.axsec / results.nunwgt
        maxwgt = results.axsec / results.nunwgt 
        tot_nevents, nb_file = 0, 0
        for result in results:  
            #misc.sprint('target:', result.axsec/result.nunwgt)
            #misc.sprint('job %s : %s %s +- %s: %s' % (result.name, result.xsec, result.axsec,\
            #                                           result.xerru, result.nunwgt))
            

            ratio = result.nunwgt/results.nunwgt
            i = result.name
            if channel.endswith(os.path.pathsep):
                path = channel[:-1] + i 
            else:
                path = channel + i
            nw = self.copy_events(fsock, pjoin(path,'events.lhe'), ratio, maxwgt)
            tot_nevents += nw
            nb_file += 1
            to_clean.append(path)
        logger.debug("Combined %s file generating %s events for %s (%.1f%%): (%fs) " , nb_file, tot_nevents, channel, 100*tot_nevents/results.nunwgt, time.time()-start)
        for path in to_clean:
            try:
                shutil.rmtree(path)
            except Exception as error:
                pass
            
    @staticmethod
    def get_fortran_str(nb):
        data = '%E' % nb
        nb, power = data.split('E')
        nb = abs(float(nb)) /10
        power = int(power) + 1
        return '%.7fE%+03i' %(nb,power)    


    def copy_events(self, fsock, input, scale_wgt, max_wgt):
        """ Copy events from separate runs into one file w/ appropriate wgts"""
        
        import collections
        wgts = collections.defaultdict(int)


        do_unweight = True
        #tmp_max_wgt = 0
        #new_wgt = self.get_fortran_str(new_wgt)
        old_line = ""
        nb_evt =0 
        skip = False
        nb_read = 0
        for line in open(input):
            if old_line.startswith("<event>"):
                nb_read +=1
                data = line.split()
                if not len(data) == 6:
                    raise MadGraph5Error("Line after <event> should have 6 entries")

                new_wgt = float(data[2]) * scale_wgt
                wgts[new_wgt] +=1
                if new_wgt < 0:
                    sign = '-'
                else:
                    sign = ''
                new_wgt = abs(new_wgt)
                skip = False
                #if new_wgt > tmp_max_wgt:
                    #misc.sprint("Found event with wgt %s higher than max wgt %s. uwgt to %s " % (new_wgt, tmp_max_wgt, max_wgt))
                #    tmp_max_wgt = new_wgt
                if do_unweight and abs(new_wgt) < random.random() * max_wgt:
                    skip = True 
                else:
                    nb_evt+=1
                    if do_unweight:
                        new_wgt = max(max_wgt, new_wgt) 

                    new_wgt = self.get_fortran_str(new_wgt)
                    line= ' %s  %s%s  %s\n' % ('   '.join(data[:2]),
                                           sign, new_wgt, '  '.join(data[3:]))
            if not skip and old_line:
                fsock.write(old_line)
            old_line = line
        if not skip and old_line:
                fsock.write(old_line) 
        #misc.sprint("Read %s events, wrote %s events: %s%%" % (nb_read, nb_evt, 100*nb_evt/nb_read if nb_read else 0))
        #misc.sprint(wgts)
        return nb_evt
    
    def get_channels(self, proc_path):
        """Opens file symfact.dat to determine all channels"""
        sympath = os.path.join(proc_path, 'symfact.dat')
        
        #ncode is number of digits needed for the bw coding
        
        ncode = int(math.log10(3)*(self.maxparticles-3))+1
        channels = []
        for line in open(sympath):
            try:
                xi, j = line.split()
            except Exception:
                break
            xi, j  = float(xi), int(j)
            
            if j > 0:
                k = int(xi) 
                npos = int(math.log10(k))+1
                #Write with correct number of digits
                if xi == k:
                    dirname = 'G%i' % k
                else:
                    dirname = 'G%.{0}f'.format(ncode) % xi
                channels.append(os.path.join(proc_path,dirname))
        return channels
    
        
        
              
